src/db_bench/db_bench.rs

```diff
@@ -0,0 +1,303 @@
+    num_rows: Option<u64>,
+    duration: Option<Duration>,
+    db: Arc<Db>,
+}
+
+impl DbBench {
+    #[allow(clippy::too_many_arguments)]
+    pub fn write(
+        key_gen_supplier: Box<dyn Fn() -> Box<dyn KeyGenerator>>,
+        val_size: usize,
+        write_options: WriteOptions,
+        write_rate: Option<u32>,
+        write_tasks: u32,
+        num_rows: Option<u64>,
+        duration: Option<Duration>,
+        db: Arc<Db>,
+    ) -> Self {
+        Self {
+            key_gen_supplier,
+            val_size,
+            write_options,
+            write_rate,
+            read_rate: None,
+            write_tasks,
+            read_tasks: 0,
+            num_rows,
+            duration,
+            db,
+        }
+    }
+
+    pub async fn run(&self) {
+        let rate_limiter = self.write_rate.map(|r| {
+            Arc::new(
+                RateLimiter::builder()
+                    .initial(r as usize)
+                    .max(r as usize)
+                    .interval(Duration::from_millis(1))
+                    .refill((r / 1000) as usize)
+                    .build(),
+            )
+        });
+        let stats_recorder = Arc::new(StatsRecorder::new());
+        let mut write_tasks = Vec::new();
+        for _ in 0..self.write_tasks {
+            let mut write_task = WriteTask::new(
+                (*self.key_gen_supplier)(),
+                self.val_size,
+                self.write_options.clone(),
+                self.num_rows,
+                self.duration,
+                rate_limiter.clone(),
+                stats_recorder.clone(),
+                self.db.clone(),
+            );
+            write_tasks.push(tokio::spawn(async move { write_task.run().await }));
+        }
+        tokio::spawn(async move { dump_stats(stats_recorder).await });
+        for write_task in write_tasks {
+            write_task.await.unwrap();
+        }
+    }
+}
+
+struct WriteTask {
+    key_generator: Box<dyn KeyGenerator>,
+    val_size: usize,
+    write_options: WriteOptions,
+    num_keys: Option<u64>,
+    duration: Option<Duration>,
+    rate_limiter: Option<Arc<RateLimiter>>,
+    stats_recorder: Arc<StatsRecorder>,
+    db: Arc<Db>,
+}
+
+impl WriteTask {
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        key_generator: Box<dyn KeyGenerator>,
+        val_size: usize,
+        write_options: WriteOptions,
+        num_keys: Option<u64>,
+        duration: Option<Duration>,
+        rate_limiter: Option<Arc<RateLimiter>>,
+        stats_recorder: Arc<StatsRecorder>,
+        db: Arc<Db>,
+    ) -> Self {
+        Self {
+            key_generator,
+            val_size,
+            write_options,
+            num_keys,
+            duration,
+            rate_limiter,
+            stats_recorder,
+            db,
+        }
+    }
+
+    async fn run(&mut self) {
+        let start = std::time::Instant::now();
+        let mut keys_written = 0u64;
+        let write_batch = 4;
```

Comment:
Curious why you introduced this. Is it just to amortize the cost of the rate limiter and stats recorder a bit more?